csharp/src/Google.Protobuf/CodedInputReader.cs

```diff
@@ -0,0 +1,788 @@
+        /// <param name="byteLength">The length of the string to be decoded, in bytes.</param>
+        /// <returns>The decoded string.</returns>
+        private string ReadStringSlow(int byteLength)
+        {
+            ThrowEndOfStreamUnless(reader.Remaining >= byteLength);
+
+            // We need to decode bytes incrementally across multiple spans.
+            int maxCharLength = CodedOutputStream.Utf8Encoding.GetMaxCharCount(byteLength);
+            char[] charArray = ArrayPool<char>.Shared.Rent(maxCharLength);
+            var decoder = CodedOutputStream.Utf8Encoding.GetDecoder();
+
+            int remainingByteLength = byteLength;
+            int initializedChars = 0;
+            while (remainingByteLength > 0)
+            {
+                int bytesRead = Math.Min(remainingByteLength, this.reader.UnreadSpan.Length);
+                remainingByteLength -= bytesRead;
+                bool flush = remainingByteLength == 0;
+
+                unsafe
+                {
+                    fixed (byte* pUnreadSpan = reader.UnreadSpan)
+                    fixed (char* pCharArray = &charArray[initializedChars])
+                    {
+                        initializedChars += decoder.GetChars(pUnreadSpan, bytesRead, pCharArray, charArray.Length - initializedChars, flush);
+                    }
+                }
+            }
+
+            string value = new string(charArray, 0, initializedChars);
+            ArrayPool<char>.Shared.Return(charArray);
+            return value;
+        }
+
+        /// <summary>
+        /// Reads an embedded message field value from the stream.
+        /// </summary>   
+        public void ReadMessage(ISpanMessage builder)
+        {
+            int length = ReadLength();
+            if (recursionDepth >= recursionLimit)
+            {
+                throw InvalidProtocolBufferException.RecursionLimitExceeded();
+            }
+
+            ++recursionDepth;
+            builder.MergeFrom(ref this);
+            CheckReadEndOfStreamTag();
+
+            --recursionDepth;
+        }
+
+        /// <summary>
+        /// Reads an embedded group field from the stream.
+        /// </summary>
+        public void ReadGroup(ISpanMessage builder)
+        {
+            if (recursionDepth >= recursionLimit)
+            {
+                throw InvalidProtocolBufferException.RecursionLimitExceeded();
+            }
+            ++recursionDepth;
+            builder.MergeFrom(ref this);
+            --recursionDepth;
+        }
+
+        /// <summary>
+        /// Reads a bytes field value from the stream.
+        /// </summary>   
+        public ByteString ReadBytes()
+        {
+            int length = ReadLength();
+
+            ThrowEndOfStreamUnless(reader.Remaining >= length);
+
+            if (length < 0)
+            {
+                throw InvalidProtocolBufferException.NegativeSize();
+            }
+
+            var data = reader.Sequence.Slice(reader.Position, length);
```

Comment:
Does it make sense to have a fast path, if `UnreadSpan.Length > length`, like in `ReadString()`?

Reply:
All that would save is a slice on a Sequence. Will think about small optimizations like that later.